hphp/runtime/vm/jit/relocation-arm.cpp

```diff
@@ -0,0 +1,993 @@
+     * range of instructions grew/shrank and therefore the internal refs
+     * may be off.
+     */
+    if (internal_refs_need_update) {
+      src = start;
+      bool ok = true;
+      while (src != end) {
+        Instruction* instr = Instruction::Cast(src);
+        TCA dest = rel.adjustedAddressAfter(src);
+        Instruction* instr2 = Instruction::Cast(dest);
+
+        // Adjust this instruction if A) it wasn't written from a pc relative
+        // instruction to an aboslute (or vice-versa) and B) it isn't a literal.
+        if (!rewrites.count(instr) && !literals.count(instr)) {
+          /*
+           * PC Relative
+           *   ADR/ADRP
+           *   LDR (literal)
+           *   B[.<cc>] (immediate)
+           */
+          if (instr->IsPCRelAddressing() ||
+              instr->IsLoadLiteral() ||
+              instr->IsCondBranchImm() ||
+              instr->IsUncondBranchImm()) {
+            TCA old_target = (TCA)instr->ImmPCOffsetTarget();
+            TCA adjusted_target = rel.adjustedAddressAfter(old_target);
+            TCA new_target = (adjusted_target) ? adjusted_target : old_target;
+
+            /*
+             * Calculate the new offset and update. We're just adjusting, and
+             * not relocating. Therefore we can't change relative instructions
+             * to absolute. Our only recourse is to mark it as far and then
+             * retry the entire relocation again.
+             */
+            int imm = Instruction::Cast(new_target) - instr2;
+            if ((instr->IsPCRelAddressing() && !is_int21(imm)) ||
+                (instr->IsLoadLiteral() && !is_int19(imm)) ||
+                (instr->IsCondBranchImm() && !is_int19(imm >> vixl::kInstructionSizeLog2)) ||
+                (instr->IsUncondBranchImm() && !is_int26(imm >> vixl::kInstructionSizeLog2))) {
+              FTRACE(3,
+                     "relocate: PC relative instruction at {} has",
+                     "internal reference 0x{:08x} which can't be adjusted.",
+                     "Will try again and far.\n",
+                     (uint64_t)instr, imm);
+              far.insert(instr);
+              ok = false;
+            } else {
+              instr2->SetImmPCOffsetTarget(Instruction::Cast(new_target));
+            }
+          }
+
+          /*
+           * Immediates
+           *   LDR (literal)
+           *   MOV/MOVK
+           */
+          if (instr->IsLoadLiteral()) {
+            TCA* addr = (TCA*)instr2->LiteralAddress();
+            TCA target = *addr;
+            TCA adjusted = rel.adjustedAddressAfter(target);
+
+            if (!adjusted) {
+              // Consider the case of a non-initialized mcprep smashableMovq
+              target = reinterpret_cast<TCA>((uint64_t(target) >> 1));
+              adjusted = rel.adjustedAddressAfter(target);
+              if (adjusted) {
+                adjusted = reinterpret_cast<TCA>((uint64_t(adjusted) << 1) | 1);
+              }
```

Comment:
Just to make sure I'm not totally off-the-mark—the reason you have to deal with this case, and literals in general, is because unlike on x64, you can't just embed your immediates within instructions, and thus the memory location in the instruction stream where the immediate is found might require position adjustment.  Is that correct?